src/ngraph_encapsulate_op.cc

```diff
@@ -179,30 +185,47 @@ class NGraphEncapsulateOp : public tf::OpKernel {
                                                         &ng_element_type));
 
       void* src_ptr = (void*)tf::DMAHelper::base(&ctx->input(i));
-      auto t = m_ng_backend->create_tensor(ng_element_type, ng_shape, src_ptr);
-
-      // Mark each tensor as non-stale if:
-      //
-      //   1. the freshness tracker says the tensor has not changed since
-      //      the last time ng_function was called, and
-      //   2. we are using the same tensor in this argument position as
-      //      the one we used last time ng_function was called.
-      if (m_freshness_tracker->IsFresh(src_ptr, ng_function) &&
-          src_ptr == last_used_src_ptrs[i]) {
-        t->set_stale(false);
+
+      auto add_input_tensor =
+          [&](shared_ptr<ngraph::runtime::TensorView>& t) -> void {
+        // Mark each tensor as non-stale if:
+        //
+        //   1. the freshness tracker says the tensor has not changed since
+        //      the last time ng_function was called, and
+        //   2. we are using the same tensor in this argument position as
+        //      the one we used last time ng_function was called.
+        if (m_freshness_tracker->IsFresh(src_ptr, ng_function) &&
+            src_ptr == last_used_src_ptrs[i]) {
+          t->set_stale(false);
+        } else {
+          t->set_stale(true);
+        }
+        last_used_src_ptrs[i] = src_ptr;
+        ng_inputs.push_back(t);
+      };
+
+      if (m_ng_backend_name == "CPU") {
+        auto t =
+            m_ng_backend->create_tensor(ng_element_type, ng_shape, src_ptr);
+        add_input_tensor(t);
       } else {
-        t->set_stale(true);
+        auto t = m_ng_backend->create_tensor(ng_element_type, ng_shape);
+        auto element_type =
+            t->get_descriptor()->get_tensor_view_layout()->get_element_type();
+        t->write(src_ptr, 0,
+                 (t->get_element_count() * element_type.bitwidth() / 8.0));
```

Comment:
@avijit-nervana @aprocter .  I will need your help in letting me know if you have a place that has this logic stored.  I'm not sure if this is the cleanest way to implement this at the Framework level.

Reply:
:thisisfine: